flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/adaptive/Executing.java

```diff
@@ -81,11 +83,15 @@ public void cancel() {
 
     @Override
     public void handleGlobalFailure(Throwable cause) {
-        handleAnyFailure(cause);
+        handleAnyFailure(null, cause);
     }
 
-    private void handleAnyFailure(Throwable cause) {
-        final FailureResult failureResult = context.howToHandleFailure(cause);
+    private void handleAnyFailure(
+            @Nullable ExecutionVertexID failingExecutionVertexId, Throwable cause) {
+        final FailureResult failureResult =
+                context.howToHandleFailure(failingExecutionVertexId, cause);
+
+        archiveExecutionFailure(failingExecutionVertexId, cause);
```

Comment:
Thinking about it once more: I guess, it's not the right location to archive the failure considering that we also want to identify concurrent failures. We haven't addressed that in this PR, yet.

To achieve that, we have to collect the failure snapshot here and pass it over to the next state (failure or restart). Any failure that pops up in these subsequent states has to be collected as well. The archiving should happen when re-instantiating the `ExecutionGraph`.

Reply:
We should also cover this in the corresponding `StateWithExecutionGraphTest` test implementations.